clusterloader2/pkg/measurement/common/network/network_performance_measurement.go

```diff
@@ -0,0 +1,449 @@
+}
+
+func (npm *networkPerformanceMeasurement) calculateMetricDataValue(dataElem *measurementutil.DataItem, metricIndex int) {
+	var aggregatePodPairMetrics []float64
+	var metricResponse []float64
+
+	npm.metricLock.Lock()
+	defer npm.metricLock.Unlock()
+
+	switch npm.podRatioType {
+	case oneToOne:
+		for _, metricResponse = range npm.metricVal {
+			if len(metricResponse) > 0 {
+				dataElem.Data[value] = metricResponse[metricIndex]
+			} else {
+				dataElem.Data[value] = 0
+			}
+		}
+		klog.Info("Metric value: ", dataElem.Data[value])
+	case manyToMany:
+		for _, metricResponse = range npm.metricVal {
+			if len(metricResponse) > 0 {
+				aggregatePodPairMetrics = append(aggregatePodPairMetrics, metricResponse[metricIndex])
+			}
+		}
+		percentile := getPercentile(aggregatePodPairMetrics)
+		dataElem.Data[perc05] = percentile.percentile05
+		dataElem.Data[perc50] = percentile.percentile50
+		dataElem.Data[perc95] = percentile.percentile95
+		klog.Info("Aggregate Metric value: ", aggregatePodPairMetrics)
+	}
+	klog.Infof("PodPairNumber: %v, Response received number: %v ", len(npm.podInfo.uniquePodPairList), len(npm.metricVal))
+}
+
+func (npm *networkPerformanceMeasurement) createResultSummary() {
+	var resultSummary testResultSummary
+	switch npm.protocol {
+	case protocolTCP:
+		npm.getMetricData(&resultSummary, tcpBandwidth, throughput)
+		resultSummary.protocol = protocolTCP
+	case protocolUDP:
+		npm.getMetricData(&resultSummary, udpPacketPerSecond, packetPerSecond)
+		npm.getMetricData(&resultSummary, udpJitter, jitter)
+		npm.getMetricData(&resultSummary, udpLatencyAverage, latency)
+		resultSummary.protocol = protocolUDP
+	case protocolHTTP:
+		npm.getMetricData(&resultSummary, httpResponseTime, responseTime)
+		resultSummary.protocol = protocolHTTP
+	}
+	resultSummary.service = "P2P"
+	resultSummary.podRatio = npm.podRatioType
+	npm.resultSummary = resultSummary
+}
+
+func (npm *networkPerformanceMeasurement) getMetricData(data *testResultSummary, metricIndex int, metricName string) {
+	var metricDataItem measurementutil.DataItem
+	metricDataItem.Data = make(map[string]float64)
+	metricDataItem.Labels = make(map[string]string)
+	metricDataItem.Labels["Metric"] = metricName
+	npm.calculateMetricDataValue(&metricDataItem, metricIndex)
+	metricDataItem.Unit = metricUnitMap[metricDataItem.Labels["Metric"]]
+	data.DataItems = append(data.DataItems, metricDataItem)
+	klog.V(3).Infof("TestResultSummary: %v", data)
+}
+
+func (wp *workerPodInfo) formUniquePodPair() {
+	nh := make(nodeHeap, len(wp.workerPodMap))
+	i := 0
+	for key, value := range wp.workerPodMap {
+		nh[i] = &workerNodeInfo{nodeName: key, numberOfPods: len(value)}
+		i++
+	}
+	heap.Init(&nh)
+	for i := 0; nh.Len() > 0; i++ {
+		firstNode := heap.Pop(&nh).(*workerNodeInfo)
+		if firstNode.numberOfPods > 1 {
+			heap.Push(&nh, &workerNodeInfo{nodeName: firstNode.nodeName, numberOfPods: firstNode.numberOfPods - 1})
```

Comment:
you can't push here, because they you can get the same node for the second node.

You first need to pop secondNode and only then push both.

Reply:
modified